### PR TITLE
gateio: update rate limit definitions

### PR DESCRIPTION
**Pull request by shazbert**
# PR Description

* Updates rate definitions for Gateio using [docs](https://www.gate.io/docs/developers/apiv4/en/#frequency-limit-rule)
* Removed func declaration `func GetRateLimit() request.RateLimitDefinitions` as this allocates a new rate limiting system for the package when SetDefaults is called. If using as a library service and multiple instances &gateio.Gateio{} are allocated and setup (for different key usage etc or other segregation management) then there can be an issue with getting 429'd. 
* Shared rate limits with websocket system for future trading integration. There are no websocket specific rate limits but you do get rate limited, see #1598 && [committed change](https://github.com/thrasher-corp/gocryptotrader/pull/1598/commits/9c56cbb413e7bca41aa42340d896cc10cb686561)

Fixes # (issue)

## Type of change

Please delete options that are not relevant and add an `x` in `[]` as item is complete.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How has this been tested

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration and
also consider improving test coverage whilst working on a certain feature or package.

- [ ] go test ./... -race
- [ ] golangci-lint run
- [ ] Test X

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally and on Github Actions with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by gloriousCode — CHANGES_REQUESTED** [review]
Thanks for updating the limits!

File: exchanges/gateio/ratelimiter.go [hunk 1]
```diff
@@ -6,49 +6,406 @@ import (
...
-	perpetualSwapPlaceOrdersRate  = 100
-	perpetualSwapPrivateRate      = 400
-	perpetualSwapCancelOrdersRate = 400
-	walletRate                    = 200
-	withdrawalRate                = 1
-
-	// interval
-	oneSecondInterval    = time.Second
-	threeSecondsInterval = time.Second * 3
+	publicTickersSpotEPL request.EndpointLimit = iota + 1
```
[the remark this answers is not in the document]
Comment (shazbert) [reply exchanges/gateio/ratelimiter.go:11]: I didn't want a value of zero for any of declared 🤷 seemed weird. 😆

File: exchanges/gateio/ratelimiter.go [hunk 2]
```diff
@@ -6,49 +6,406 @@ import (
...
+	optionsSubmitOrderEPL
+	optionsOrdersEPL
+	optionsCancelOrdersEPL
+	optionsOrderEPL
+	optionsCancelOrderEPL
+	optionsTradingHistoryEPL
 )
 
-// GetRateLimit returns the rate limiter for the exchange
-func GetRateLimit() request.RateLimitDefinitions {
```
[the remark this answers is not in the document]
Comment (shazbert) [reply exchanges/gateio/ratelimiter.go:41]: Yeah I was considering that but wanted to go more in depth in another PR for a `request.RegisterRateLimit(exchangeStructPointer any, map[request.EndpointLimit]*request.RateLimiter) func` and maybe a protected type just for exchanges. That way any future exchange package implementations don't reintroduce the whoopsie again. 

1) I can add your changes to this PR for now or
2) You can open up a PR against this and I can merge or
3) You can open a PR against master once this is merged cause I am focused on other things at the moment. 

let me know how you want to proceed?

File: exchanges/gateio/ratelimiter.go [hunk 3]
```diff
@@ -6,49 +6,406 @@ import (
...
+	walletSubAccountToSubAccountTransferEPL: personalAccountRateLimit(),
+	walletWithdrawStatusEPL:                 standardRateLimit(),
+	walletSubAccountBalancesEPL:             personalAccountRateLimit(),
+	walletSubAccountMarginBalancesEPL:       personalAccountRateLimit(),
+	walletSubAccountFuturesBalancesEPL:      personalAccountRateLimit(),
+	walletSubAccountCrossMarginBalancesEPL:  personalAccountRateLimit(),
+	walletSavedAddressesEPL:                 standardRateLimit(),
+	walletTradingFeeEPL:                     standardRateLimit(),
+	walletTotalBalanceEPL:                   personalAccountRateLimit(),
+	walletWithdrawEPL:                       request.NewRateLimitWithWeight(time.Second*3, 1, 1), // 1r/3s
```
[the remark this answers is not in the document]
Comment (shazbert) [reply exchanges/gateio/ratelimiter.go:257]: This was the only single one in docs for 1 request every 3 seconds only on withdrawal from wallet. 🤷

**Review by gloriousCode — APPROVED** [review]
tACK!

File: exchanges/gateio/ratelimiter.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (gloriousCode) [reply exchanges/gateio/ratelimiter.go:41]: I think its something to roll out separately in a different PR. Like I'm okay without it, as per my above snippet. Can be revisited in future and per my second snippet, isn't crazy complex to implement

**Review by thrasher- — APPROVED** [review]
utACK, rate limit EPLs match expected endpoints; nice work on this as that is always tedious

**Pull request merged**